sympy/solvers/ode/systems.py

```diff
@@ -408,251 +409,382 @@ def jordan_chains(A):
-    # Some parts of code is repeated, this needs to be taken care of
-    # The constant vector obtained here can be done so in the match
-    # function itself.
-    eq = match_['eq']
-    func = match_['func']
-    fc = match_['func_coeff']
-    n = len(eq)
-    t = list(list(eq[0].atoms(Derivative))[0].atoms(Symbol))[0]
-    constants = numbered_symbols(prefix='C', cls=Symbol, start=1)
 
-    # This needs to be modified in future so that fc is only of type Matrix
-    M = -fc if type(fc) is Matrix else Matrix(n, n, lambda i,j:-fc[i,func[j],0])
+    if not isinstance(A, MatrixBase):
+        raise ValueError(filldedent('''\
+            The coefficients of the system of ODEs should be of type Matrix
+        '''))
+
+    if not A.is_square:
+        raise NonSquareMatrixError(filldedent('''\
+            The coefficient matrix must be a square
+        '''))
+
+    if b is not None:
+        if not isinstance(b, MatrixBase):
+            raise ValueError(filldedent('''\
+                The non-homogeneous terms of the system of ODEs should be of type Matrix
+            '''))
+
+        if A.rows != b.rows:
+            raise ValueError(filldedent('''\
+                The system of ODEs should have the same number of non-homogeneous terms and the number of
+                equations
+            '''))
+
+    if B is not None:
+        if not isinstance(B, MatrixBase):
+            raise ValueError(filldedent('''\
+                The antiderivative of coefficients of the system of ODEs should be of type Matrix
+            '''))
+
+        if not B.is_square:
+            raise NonSquareMatrixError(filldedent('''\
+                The antiderivative of the coefficient matrix must be a square
+            '''))
+
+        if A.rows != B.rows:
+            raise ValueError(filldedent('''\
+                        The coefficient matrix and its antiderivative should have same dimensions
+                    '''))
+
+    if not any(type == "type{}".format(i) for i in range(1, 5)):
+        raise ValueError(filldedent('''\
+                    The input to type should be a valid one
+                '''))
+
+    n = A.rows
 
+    constants = numbered_symbols(prefix='C', cls=Symbol, start=1)
     Cvect = Matrix(list(next(constants) for _ in range(n)))
 
-    # The code in if block will be removed when it is made sure
-    # that the code works without the statements in if block.
-    if "commutative_antiderivative" not in match_:
-        B, is_commuting = _is_commutative_anti_derivative(M, t)
+    if (type == "type2" or type == "type4") and b is None:
+        b = zeros(n, 1)
 
-        # This course is subject to change
-        if not is_commuting:
-            return None
+    if type == "type1" or type == "type2":
+        P, J = matrix_exp_jordan_form(A, t)
+        P = simplify(P)
+
+        sol_vector = P * (J * Cvect) if type == "type1" else P * J * ((J.inv() * P.inv() * b).applyfunc(
+            lambda x: Integral(x, t)) + Cvect)
```

Comment:
It's better not to use inline if/else unless the expressions are very short. This should be 
```
if type == "type1":
   ...
elif ...
```